apps/designer/app/designer/shared/design-tokens-manager/design-tokens-manager.tsx

```diff
@@ -0,0 +1,178 @@
+import {
+  Button,
+  Flex,
+  InputErrorsTooltip,
+  Label,
+  Popover,
+  PopoverContent,
+  PopoverHeader,
+  PopoverPortal,
+  PopoverTrigger,
+  TextArea,
+  TextField,
+} from "@webstudio-is/design-system";
+import { PlusIcon } from "@webstudio-is/icons";
+import { type FormEvent, useState, useEffect } from "react";
+import { useDesignTokens } from "~/shared/nano-states";
+import type { Publish } from "~/shared/pubsub";
+import { CollapsibleSection } from "../inspector";
+import { groups } from "./groups";
+import type { DesignToken } from "./schema";
+import { filterByType, findByName } from "./utils";
+
+declare module "~/shared/pubsub" {
+  export interface PubsubMap {
+    updateTokens: Array<DesignToken>;
+  }
+}
+
+const validate = (
+  tokens: Array<DesignToken>,
+  data: Partial<DesignToken>
+): { name: Array<string>; value: Array<string>; hasErrors: boolean } => {
+  const name = [];
+  const value = [];
+
+  if (String(data.name).trim() === "") name.push("Name is required");
+  if (findByName(tokens, data?.name)) name.push("Name is already taken");
+  if (String(data.value).trim() === "") value.push("Value is required");
+
+  return {
+    name,
+    value,
+    hasErrors: name.length !== 0 || value.length !== 0,
+  };
+};
+
+const initialErrors = {
+  name: [],
+  value: [],
+  hasErrors: false,
+};
+
+const TokenEditor = ({
+  group,
+  name,
+  type,
+  onChangeComplete,
+}: {
+  group: string;
+  type: string;
+  name?: string;
+  onChangeComplete: (token: DesignToken) => void;
+}) => {
+  const [tokens] = useDesignTokens();
+  const [isOpen, setIsOpen] = useState(false);
+  const [errors, setErrors] =
+    useState<ReturnType<typeof validate>>(initialErrors);
+  const isNew = name === undefined;
+
+  useEffect(() => {
+    if (isOpen === false && errors.hasErrors) {
+      setErrors(initialErrors);
+    }
+  }, [isOpen]);
+
+  const handleChangeOrSubmit = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    const formData = new FormData(event.currentTarget);
+    const data = Object.fromEntries(formData);
+    const errors = validate(tokens, data);
+
+    if (errors.hasErrors && event.type === "submit") {
+      setErrors(errors);
+    }
+
+    if (errors.hasErrors === false && event.type === "submit") {
+      onChangeComplete({ ...data, group, type } as DesignToken);
+      setIsOpen(false);
+    }
+  };
```

Comment:
This function seem to do anything useful only when `event.type === "submit"`. Should we even call it on change?

Reply:
rewrote